Add testing docs: advanced scenarios and CI/CD patterns (#456)
microsoft/aspire.dev: src/frontend/src/content/docs/testing/testing-in-ci.mdx

    await app.ResourceNotifications.WaitForResourceHealthyAsync(
        "webfrontend", cts.Token);

    using var httpClient = app.CreateHttpClient("webfrontend");
    using var response = await httpClient.GetAsync("/");
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
}
```

## Configure Azure authentication in CI

When running tests that involve Azure resources (such as Azure Cosmos DB, Azure Service Bus, or Azure Storage), you must configure Azure authentication appropriately for your CI environment.

### Local vs. CI authentication

In local development, Aspire uses your developer identity (via `DefaultAzureCredential`, which tries Visual Studio, Azure CLI, and other sources). In CI, no developer identity is available, so you must configure a **service principal** or **managed identity**.

A common symptom of misconfigured CI authentication is errors like:

```
The principal type 'User' is not allowed. Expected 'ServicePrincipal'.
```

This occurs when a role assignment in your Aspire AppHost is configured with `principalType: "User"` but the CI pipeline is running as a service principal.

### Use `DefaultAzureCredential` with environment variables

> The recommended approach is to configure a service principal and set the standard Azure SDK environment variables in your CI pipeline. `DefaultAzureCredential` automatically picks up these environment variables:

Reviewer: Do we need to link to the new content that addressed this issue https://github.com/microsoft/aspire.dev/pull/424?

| Environment variable | Description |
|----------------------|-------------|
| `AZURE_CLIENT_ID` | The application (client) ID of your service principal |
| `AZURE_TENANT_ID` | Your Azure Active Directory tenant ID |
| `AZURE_CLIENT_SECRET` | The client secret of your service principal |

> <Tabs>

Reviewer: Please always use a syncKey on all Tabs.

```csharp title="C# — IntegrationTest.cs"
var appHost = await DistributedApplicationTestingBuilder
    .CreateAsync<Projects.MyAppHost>(
    [
        "DcpPublisher:RandomizePorts=true"
    ]);
```

<LearnMore>
For information about disabling port randomization, see [Testing overview: Disable port randomization](/testing/overview/#disable-port-randomization).
</LearnMore>

## Example CI workflow

The following is a complete GitHub Actions workflow for running Aspire integration tests:

<Aside type="note">
GitHub Actions automatically sets the `CI` environment variable to `true` for all workflow runs. The test code from the earlier [Use longer timeouts in CI](#use-longer-timeouts-in-ci) section checks for this variable to apply extended timeouts. Other CI platforms (Azure DevOps, CircleCI, and others) also set `CI=true` by default.
</Aside>

```yaml title="YAML — .github/workflows/integration-tests.yml"
name: Integration Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4

      - name: Setup .NET
        uses: actions/setup-dotnet@v4
        with:

>           dotnet-version: '9.x'

Reviewer: ```suggestion
          dotnet-version: '10.x'
```

---
title: Testing in CI/CD pipelines
description: Learn how to run Aspire integration tests reliably in CI/CD environments, such as GitHub Actions and Azure DevOps, covering timeout configuration, Azure authentication, and container requirements.
---

import { Aside, Tabs, TabItem } from '@astrojs/starlight/components';
import LearnMore from '@components/LearnMore.astro';

This article covers how to run Aspire integration tests reliably in continuous integration (CI) and continuous deployment (CD) environments, addressing common challenges such as test timeouts, Azure authentication, and container requirements.

## Container requirements

> Aspire integration tests typically start containers (databases, caches, and other services) as part of the test run. Your CI environment must have a container runtime available.

Reviewer 2: ```suggestion
Aspire integration tests typically start containers, such as databases, caches, and other services, as part of the test run. Your CI environment must have a container runtime available.
```

> - **GitHub Actions**: For Linux-based Aspire test containers, use `ubuntu-*` runners, which have Docker available by default. GitHub-hosted `windows-*` and `macos-*` runners do not provide a Docker Engine suitable for running Linux containers; use self-hosted runners with a configured container runtime if you must target those operating systems.

Reviewer 2: I don't think "engine" should be capitalized. Also "if you must" makes it sound like a bad thing to do:

```suggestion
- **GitHub Actions**: For Linux-based Aspire test containers, use `ubuntu-*` runners, which have Docker available by default. GitHub-hosted `windows-*` and `macos-*` runners do not provide a Docker engine suitable for running Linux containers; use self-hosted runners with a configured container runtime if you need to target those operating systems.
```

- **Azure DevOps**: Docker is available on Microsoft-hosted agents (`ubuntu-latest`, `windows-latest`). Ensure the agent pool supports Docker.

> - **Self-hosted runners**: Install and start Docker or another compatible container runtime (such as Podman) before running tests.

Reviewer 2: ```suggestion
- **Self-hosted runners**: Install and start Docker or another compatible container runtime, such as Podman, before running tests.
```

<Aside type="caution">
Some CI environments restrict container networking, privileged containers, or certain Docker features. If you encounter container startup failures in CI, check your runner's Docker capabilities and network configuration.
</Aside>

## Configure timeouts to prevent hanging tests

> One of the most common issues when running Aspire tests in CI is tests hanging indefinitely because resources never reach their expected state. Always configure explicit timeouts for resource waiting and test execution.

Reviewer 2: ```suggestion
One of the most common issues experienced when running Aspire tests in CI is a test that hangs indefinitely because resources never reach their expected state. Always configure explicit timeouts for resource waiting and test execution.
```

### Set timeouts on resource waiting

The `WaitForResourceAsync` and `WaitForResourceHealthyAsync` methods accept a `CancellationToken`. Always pass a token with a timeout to avoid indefinite waits:

```csharp title="C# — IntegrationTest.cs"
using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(5));

await app.ResourceNotifications.WaitForResourceHealthyAsync(
    "webfrontend",
    cts.Token);
```

### Set timeouts on `BuildAsync` and `StartAsync`

> Use `WaitAsync` to apply a timeout to the build and startup operations:

Reviewer 2: ```suggestion
Use `WaitAsync` to apply a timeout to the `BuildAsync` and `StartAsync` operations:
```

CI environments are often slower than local development machines due to network latency when pulling container images, limited CPU and memory, and parallel test execution. Consider using environment-aware timeouts:

```csharp title="C# — IntegrationTest.cs"
private static readonly TimeSpan DefaultTimeout = 
    Environment.GetEnvironmentVariable("CI") is not null
        ? TimeSpan.FromMinutes(5)   // Longer timeout in CI
        : TimeSpan.FromSeconds(30); // Shorter timeout locally

[Fact]
public async Task GetWebResourceRootReturnsOkStatusCode()
{
    var appHost = await DistributedApplicationTestingBuilder
        .CreateAsync<Projects.MyAppHost>();

    await using var app = await appHost.BuildAsync()
        .WaitAsync(DefaultTimeout);
    await app.StartAsync().WaitAsync(DefaultTimeout);

    using var cts = new CancellationTokenSource(DefaultTimeout);
    await app.ResourceNotifications.WaitForResourceHealthyAsync(
        "webfrontend", cts.Token);

    using var httpClient = app.CreateHttpClient("webfrontend");
    using var response = await httpClient.GetAsync("/");
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
}
```

## Configure Azure authentication in CI

> When running tests that involve Azure resources (such as Azure Cosmos DB, Azure Service Bus, or Azure Storage), you must configure Azure authentication appropriately for your CI environment.

Reviewer 2: ```suggestion
When running tests that involve Azure resources, such as Azure Cosmos DB, Azure Service Bus, or Azure Storage, you must configure Azure authentication appropriately for your CI environment.
```

> ### Local vs. CI authentication

Reviewer 2: ```suggestion
### Local versus CI authentication
```

> In local development, Aspire uses your developer identity (via `DefaultAzureCredential`, which tries Visual Studio, Azure CLI, and other sources). In CI, no developer identity is available, so you must configure a **service principal** or **managed identity**.

Reviewer 2: I don't see the point of using brackets here.
```suggestion
In local development, Aspire uses your developer identity via `DefaultAzureCredential`, which tries Visual Studio, Azure CLI, and other sources. In CI, no developer identity is available, so you must configure a **service principal** or **managed identity**.
```

> A common symptom of misconfigured CI authentication is errors like:

Reviewer 2: The mix of singular and plural is grammatically correct but sounds awkward. Why not use:
```suggestion
A common symptom of misconfigured CI authentication is an error like:
```

```csharp title="C# — IntegrationTest.cs"
[Fact]
public async Task TestWithAzureCosmosDb()
{
    if (Environment.GetEnvironmentVariable("AZURE_CLIENT_ID") is null
        || Environment.GetEnvironmentVariable("AZURE_SUBSCRIPTION_ID") is null)
    {
        // Skip when Azure credentials are not available
        throw new Xunit.Sdk.SkipException("""
            Azure credentials not configured; skipping test.
            """);
    }

    var appHost = await DistributedApplicationTestingBuilder
        .CreateAsync<Projects.MyAppHost>();

    // ... rest of test
}
```

<Aside type="tip">
Consider using test categories or traits to separate tests that require Azure credentials from tests that run entirely on local containers. This lets CI pipelines choose which tests to run based on available credentials.
</Aside>

## Run tests in parallel

Aspire tests use random port assignment by default, which allows multiple test instances to run concurrently without port conflicts. This is controlled by the `DcpPublisher:RandomizePorts` setting, which is enabled by default in the testing builder.

> For CI environments running multiple test classes in parallel, random ports help prevent failures caused by port collisions. If you've disabled random ports (for example to match a specific port in a health check URL), re-enable them for CI:

Reviewer 2: ```suggestion
For CI environments running multiple test classes in parallel, random ports help prevent failures caused by port collisions. If you've disabled random ports, for example to match a specific port in a health check URL, re-enable them for CI:
```

```csharp title="C# — IntegrationTest.cs"
var appHost = await DistributedApplicationTestingBuilder
    .CreateAsync<Projects.MyAppHost>(
    [
        "DcpPublisher:RandomizePorts=true"
    ]);
```

<LearnMore>
For information about disabling port randomization, see [Testing overview: Disable port randomization](/testing/overview/#disable-port-randomization).
</LearnMore>

## Example CI workflow

The following is a complete GitHub Actions workflow for running Aspire integration tests:

<Aside type="note">

> GitHub Actions automatically sets the `CI` environment variable to `true` for all workflow runs. The test code from the earlier [Use longer timeouts in CI](#use-longer-timeouts-in-ci) section checks for this variable to apply extended timeouts. Other CI platforms (Azure DevOps, CircleCI, and others) also set `CI=true` by default.

Reviewer 2: ```suggestion
GitHub Actions automatically sets the `CI` environment variable to `true` for all workflow runs. The test code from the earlier [Use longer timeouts in CI](#use-longer-timeouts-in-ci) section checks for this variable to apply extended timeouts. Other CI platforms such as Azure DevOps, CircleCI, and others, also set `CI=true` by default.
```
